Merge Tally counter / wind speed into develop

aufdenkampe opened the pull request:
This is for  the "Tally" high-speed, ultra-low-power counting device that @bschulz1701  created.
https://github.com/NorthernWidget-Skunkworks/Project-Tally​
https://github.com/NorthernWidget-Skunkworks/Tally_Library/tree/Dev_I2C

I'm using it with the Inspeed Version II Reed-Switch Anemometer
https://www.store.inspeed.com/Inspeed-Version-II-Reed-Switch-Anemometer-Sensor-Only-WS2R.htm

It all works well enough to deploy, but when I have debugging on it says it's not setup properly. @SRGDamia1 & @bschulz1701 , let me know if there are any additional tweaks that might improve my code wrappers.

Here's the debug output that repeatedly says that the Tally `was never properly set up, attempting setup now!`:
```
------------------------------------------
Clearing value array for TallyCounterI2C at I2C_0x33 <--SensorBase
TallyCounterI2C at I2C_0x33 was never properly set up, attempting setup now! <--SensorBase
Setting up TallyCounterI2C attached at I2C_0x33 which can return up to 1 variable[s]. <--SensorBase
It warms up in 0 ms, is stable after 0 ms, and takes a single measurement in 0 ms. <--SensorBase
1 individual measurements will be averaged for each reading. <--SensorBase
Starting measurement on TallyCounterI2C at I2C_0x33 <--SensorBase
TallyCounterI2C at I2C_0x33 is reporting: <--TallyCounterI2C
  Events: 6 <--TallyCounterI2C
Putting 6.00 in result array for variable 0 from TallyCounterI2C at I2C_0x33 <--SensorBase
Averaging results from TallyCounterI2C at I2C_0x33 over 1 reading[s] <--SensorBase
    ->Result # 0 : 6.00 <--SensorBase

Notifying variables registered to TallyCounterI2C at I2C_0x33 of value update. <--SensorBase
Sending value update from TallyCounterI2C at I2C_0x33 to variable 0 which is counter ... <--SensorBase
... received 6.00 <--VariableBase
Current logger time is 2019-06-27T15:53:10-05:00
-----------------------
TallyCounterI2C at I2C_0x33 reports counter is 6 event
-----------------------
```

aufdenkampe commented:
From @bschulz1701, regarding updates to https://github.com/NorthernWidget-Skunkworks/Tally_Library/tree/Dev_I2C:

> By the way, not sure how tight your power budget is, but I just added the sleep feature into the software (it is supported by the firmware you have already)
> 
> I made an update to the library and the read example, but once you pull the updated library, you just need to call
> `Counter.Sleep();`
> After you call `begin` and it will operate in a sleep/doze mode, automatically waking and going back to sleep for I2C messages.
> 
> And if you really want to minimize power, you can just connect it to a switched rail as well and the on board capacitor will run it when the logger is off, then recharge when you switch back on for your logging cycle.
> 
> Power Consumption:
> No Sleep: ~4mA @ 3.3v
> Sleep: <100uA @ 3.3v
> Switched: N/A

SRGDamia1 commented on src/sensors/TallyCounterI2C.cpp:
```diff
@@ -0,0 +1,123 @@
+
+    // Restart always needed after power-up to set sampling modes
+    counter_internal.begin(_i2cAddressHex);
+
```
Is this needed?  It seems like all the Tally's "begin" function is doing is starting up wire.  I suspect you don't need a wake function at all.

aufdenkampe replied:
Removed begin in wake with f0823303. Should I delete the entire wake function, or just leave this barebones version?

SRGDamia1 commented on src/sensors/TallyCounterI2C.cpp:
```diff
@@ -0,0 +1,123 @@
+TallyCounterI2C::TallyCounterI2C(uint8_t i2cAddressHex)
+     : Sensor("TallyCounterI2C", TALLY_NUM_VARIABLES,
+              TALLY_WARM_UP_TIME_MS, TALLY_STABILIZATION_TIME_MS, TALLY_MEASUREMENT_TIME_MS,
+              -1, -1, 1)
```
Can the sensor ever be powered down?

aufdenkampe replied:
Generally counting devices such as this and the [TippingBucketRainCounter](https://github.com/EnviroDIY/TippingBucketRainCounter) need to always be "on", so that they can count an event. This is why I set `powerPin = -1` and `measurementsToAverage = 1`. 

However, @bschulz1701 designed the Tally device with a super capacitor, so once that is charged it can be powered down between readings.

I'll add the powerPin as an option.

aufdenkampe replied:
Fixed with commits 3f8c45df and f0823303

SRGDamia1 commented on src/sensors/TallyCounterI2C.cpp:
```diff
@@ -0,0 +1,123 @@
+
+    return true;
+}
+
```
Need to create the sleep function.

aufdenkampe replied:
Addressed in 06776b01. Note that the Tally has an automatic sleep mode between event counts, which is enabled in the setup/begin function. There is no external function to put the sensor to sleep.

SRGDamia1 commented on src/sensors/TallyCounterI2C.cpp:
```diff
@@ -0,0 +1,123 @@
+
+    // Get data from external event counter
+    // // if the 'requestFrom' returns 0, it means no bytes were received
+    // if (Wire.requestFrom(int(_i2cAddressHex), 2))
```
Does the conditional not work?

aufdenkampe replied:
It didn't work, but your fix with 5a972b5f40972eec05b4b278ac9264f63ced1acc did work. Thanks.

SRGDamia1 commented on src/sensors/TallyCounterI2C.h:
```diff
@@ -0,0 +1,90 @@
+public:
+    // The constructor - all arguments are optional
+    // Address of I2C device is 0x08 by default
+    // Depth of rain per tip event in mm is 0.2mm by default
```
fix comments

aufdenkampe replied:
Done by you with https://github.com/EnviroDIY/ModularSensors/pull/269/commits/5a972b5f40972eec05b4b278ac9264f63ced1acc

aufdenkampe commented:
@SRGDamia1, thanks for this review and all your comments. There's definitely a bit more work and testing to do. I'll get back to this early next week, after I've recovered from the back-from-vacation pile of things to do.

aufdenkampe commented:
@SRGDamia1, I can't get `sensor_tests/Tally_simple_logging` to compile, but I'm not sure if I still have the `platformio.ini` set up right for PIO 4.0.

This is what I get:
```
...
Linking .pio/build/mayfly/firmware.elf
/var/folders/tx/14mbtmrs0rq2b6mkfxdh54r00000gp/T//ccrq7yu8.ltrans0.ltrans.o: In function `calculateWindSpeed()':
<artificial>:(.text+0x432): undefined reference to `Variable::getValue(bool)'
/var/folders/tx/14mbtmrs0rq2b6mkfxdh54r00000gp/T//ccrq7yu8.ltrans0.ltrans.o: In function `global constructors keyed to 65535_0_Tally_simple_logging.ino.cpp.o.2457':
<artificial>:(.text.startup+0x78): undefined reference to `ProcessorStats::ProcessorStats(char const*)'
<artificial>:(.text.startup+0x82): undefined reference to `MaximDS3231::MaximDS3231(unsigned char)'
<artificial>:(.text.startup+0x8c): undefined reference to `TallyCounterI2C::TallyCounterI2C(unsigned char)'
<artificial>:(.text.startup+0xc0): undefined reference to `Variable::Variable(Sensor*, unsigned char, unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0xee): undefined reference to `Variable::Variable(float (*)(), unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0x128): undefined reference to `Variable::Variable(Sensor*, unsigned char, unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0x15a): undefined reference to `Variable::Variable(Sensor*, unsigned char, unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0x188): undefined reference to `Variable::Variable(Sensor*, unsigned char, unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0x1b6): undefined reference to `Variable::Variable(Sensor*, unsigned char, unsigned char, char const*, char const*, char const*, char const*)'
<artificial>:(.text.startup+0x1da): undefined reference to `VariableArray::VariableArray()'
<artificial>:(.text.startup+0x1fa): undefined reference to `Logger::Logger()'
/var/folders/tx/14mbtmrs0rq2b6mkfxdh54r00000gp/T//ccrq7yu8.ltrans0.ltrans.o: In function `main':
<artificial>:(.text.startup+0x464): undefined reference to `Logger::setLoggerTimeZone(signed char)'
<artificial>:(.text.startup+0x46a): undefined reference to `Logger::setRTCTimeZone(signed char)'
<artificial>:(.text.startup+0x47e): undefined reference to `Logger::setLoggerPins(signed char, signed char, signed char, signed char, signed char)'
<artificial>:(.text.startup+0x48c): undefined reference to `VariableArray::begin(unsigned char, Variable**)'
<artificial>:(.text.startup+0x4a0): undefined reference to `Logger::begin(char const*, unsigned int, VariableArray*)'
<artificial>:(.text.startup+0x4b8): undefined reference to `VariableArray::setupSensors()'
<artificial>:(.text.startup+0x4c2): undefined reference to `Logger::createLogFile(bool)'
<artificial>:(.text.startup+0x4ca): undefined reference to `Logger::systemSleep()'
<artificial>:(.text.startup+0x4d2): undefined reference to `Logger::logData()'
/var/folders/tx/14mbtmrs0rq2b6mkfxdh54r00000gp/T//ccrq7yu8.ltrans0.ltrans.o: In function `_GLOBAL__sub_D_pcmsk':
<artificial>:(.text.exit+0x4): undefined reference to `Logger::~Logger()'
<artificial>:(.text.exit+0xc): undefined reference to `VariableArray::~VariableArray()'
<artificial>:(.text.exit+0x14): undefined reference to `TallyCounterI2C::~TallyCounterI2C()'
<artificial>:(.text.exit+0x1c): undefined reference to `MaximDS3231::~MaximDS3231()'
<artificial>:(.text.exit+0x24): undefined reference to `ProcessorStats::~ProcessorStats()'
collect2: error: ld returned 1 exit status
*** [.pio/build/mayfly/firmware.elf] Error 1
========================== [ERROR] Took 6.66 seconds ==========================
```

aufdenkampe commented:
Update. 
I changed the way I set up my `platformio.ini` file:
- Under `[platformio]`, removed `include_dir = src`
- Under `[env:mayfly]`, replaced `lib_extra_dirs = . ` with `lib_extra_dirs = src`

Unfortunately, for _any_ of the examples I now get this error:
```
/Users/aaufdenkampe/Documents/Arduino/EnviroDIY_ModularSensors/examples/simple_logging/simple_logging.ino:44:36: fatal error: sensors/ProcessorStats.h: No such file or directory
compilation terminated.
```
I can't figure that one out.

SRGDamia1 commented:
It's just picking on ProcessorStat's because that's the first file from the library called.

You need to add ```.``` (meaning the current project folder) to your ```lib_extra_dirs``` directory, not ```src```.  It's searching for the library files in a```src``` subdirectory of the ```lib_extra_dirs``` directory, so it doesn't find the library files because there in ./src not ./src/src.  Then you need to explicitly _ignore_ all of the other library containing folders so they don't get compiled twice (which would also cause issues):
```ini
lib_ignore =
    .git
    .pio
    .vscode
    doc
    examples
    sensor_tests
```

https://community.platformio.org/t/finding-libraries-in-subfolders-submodules-on-git/1508/14

aufdenkampe commented:
@SRGDamia1, Thank you!
I had tried `lib_extra_dirs = .` and also `lib_extra_dirs = . ,src`,  but I did not also implement the `lib_ignore stuff` *at the same time*. I guess that’s what I get for trial and error. I must say that reading the PlatformIO documentation wasn’t super helpful either, because a lot of it was over my head.

I'll get back to this Tally testing later today.

aufdenkampe commented:
Everything appears to be working after making fixes suggested by @SRGDamia1 in https://github.com/EnviroDIY/ModularSensors/pull/269#pullrequestreview-256606845.

Next step for merging into develop is to update this branch to develop

aufdenkampe commented:
Everything is working with ModularSensors v0.24.5, including sensor testing mode and debugging.

@SRGDamia1 @ @bschulz1701, can you suggest a way to internally track the time interval between readings of the "count", either in the Tally_Library or in the ModularSensors library source files? If this could be easily done, then we could have "counts per second" as a measurement output from the Tally. This would allow us to skip setting up Calculated Variables in the Modular Sensors code, and it would also give the right answer for Wind Speed when in sensor testing mode.

@SRGDamia1, also, I was having trouble entering debugging modes from PlatformIO using for example:
```ini
build_flags =
    -DMS_VARIABLEARRAY_DEBUG
```
Did this syntax change? I could debug without problem if I uncommented the define in the header files. If I used PlatformIO, I would get this compile error:

```
Compiling .pio/build/mayfly/FrameworkArduino/wiring_shift.c.o
Archiving .pio/build/mayfly/libFrameworkArduino.a
Indexing .pio/build/mayfly/libFrameworkArduino.a

Linking .pio/build/mayfly/firmware.elf
/var/folders/tx/14mbtmrs0rq2b6mkfxdh54r00000gp/T//ccUqWlVA.ltrans0.ltrans.o: In function `main':
<artificial>:(.text.startup+0x23a): undefined reference to `setup'
<artificial>:(.text.startup+0x23e): undefined reference to `loop'
collect2: error: ld returned 1 exit status
*** [.pio/build/mayfly/firmware.elf] Error 1
```

SRGDamia1 commented:
The debug syntax hasn't changed.  Did you try just re-running the compiler?  I get that error with missing set up and loop often when I first compile something and it usually magically resolves itself the second try.  I haven't ever taken the time to sort out the cause.

aufdenkampe commented:
@SRGDamia1, you have reviewed this previously. Are you interested in looking it over once more before I merge into develop and then to master to issue another release? 

If you don't see a need, I'm happy to move it forward. I've recently tested it and it works well.
